src/internal/group_api/mod.rs

```diff
@@ -526,6 +530,82 @@ pub async fn group_create<CR: rand::CryptoRng + rand::RngCore>(
     resp.try_into()
 }
 
+#[derive(Debug, Clone)]
+pub struct GroupUpdatePrivateKeyResult {
+    needs_rotation: bool,
+}
+
+impl GroupUpdatePrivateKeyResult {
+    /// True if this group's private key requires additional rotation
+    pub fn needs_rotation(&self) -> bool {
+        self.needs_rotation
+    }
+}
+
+/// Rotate the group's private key. The public key for the group remains unchanged.
+pub async fn group_rotate_private_key<CR: rand::CryptoRng + rand::RngCore>(
+    recrypt: &Recrypt<Sha256, Ed25519, RandomBytes<CR>>,
+    group_id: &GroupId,
+    auth: &RequestAuth,
+    device_private_key: &PrivateKey,
+) -> Result<GroupUpdatePrivateKeyResult, IronOxideErr> {
+    let group_info = group_get_request(&auth, group_id).await?;
+    let admins = match &group_info.admin_ids {
+        Some(user_id_strings) => user_id_strings
+            .into_iter()
+            .map(|string_id| string_id.to_owned().try_into())
+            .collect::<Result<Vec<_>, _>>(),
+        None => Err(IronOxideErr::NotGroupAdmin(group_id.to_owned())),
+    }?;
+    let (_, admins_with_keys) = get_user_keys(auth, &admins).await?;
+    let encrypted_group_key = group_info
+        .encrypted_private_key
+        .ok_or(IronOxideErr::NotGroupAdmin(group_id.to_owned()))?;
```

Comment:
It seems like this check should be the first thing we do in this method. It's the quickest tell to know if the user is an admin of the group and therefore has the ability to do the rotation. As it is now, you're going to get the public keys for all of the admins before you know the the user can even rotate the key.

Reply:
This is roughly what Tonia is doing in her PR as well. We only return the encrypted group private key if the user is an admin, so it's the best way we have right now to check that.